In [1]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [ ]:
# Load the given datasets
dfc = pd.read_csv('./chronic_kidney_disease_categorical.csv')
dfn = pd.read_csv('./chronic_kidney_disease_numerical.csv')

# Print the data
print(dfc)
print(dfn)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: 
Each feature represents a certain attribute of the data set. Together, these features inform the label, or the outcome of what we're trying to predict - does the patient have CKD or not. There are categorical and numerical features in the dataset. 
Examples of numerical columns: bp, bgr, and bu. 
Examples of categorical columns: al, sg, su.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [ ]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
nDuplicates = dfn.duplicated(subset = "unique_id", keep = False)
print((nDuplicates).sum())

# Drop these duplicate rows
dfnClean = dfn.drop_duplicates(subset = "unique_id", keep = 'first')

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cDuplicates = dfc.duplicated(subset = "unique_id", keep = False)
#print(cDuplicates.sum())
print(cDuplicates)
dfcClean = dfc.drop_duplicates(subset = "unique_id", keep = 'first')
# print(len(dfcClean))
# print(len(dfnClean))

59
0       True
1      False
2      False
3      False
4      False
       ...  
425     True
426     True
427     True
428     True
429     True
Length: 430, dtype: bool


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: Not all unique ids are present in both datasets because when I looked for the number of rows in each cleaned dataset (no unique_id is repeated), they have different values. The categorical dataset has 400 entries while the numerical one has 390 entries. Since I'm joining dfnClean with dfcClean, unique_ids that only appear in dfnClean and not dfcClean will show up with NaN in those columns part of dfcClean. I'm performing an outer join.

In [ ]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
mergedDf = pd.merge(dfnClean, dfcClean, on = "unique_id", how = "outer")
#Print the combined dataset
print(mergedDf)
print(len(mergedDf))
# print(len(dfcClean))
# print(len(dfnClean))

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [ ]:
# Calculate the percentage of rows that contain atleast one missing value
mergedDf['missingVal'] = mergedDf.isna().any(axis = 1)
missingValPercentage = ((mergedDf['missingVal'].sum()/len(mergedDf))) * 100 
#can I add another column? and then delete afterwards?

# Print %
print((missingValPercentage))

# Drop these rows from the dataset
cond = mergedDf['missingVal']
mergedDfCleaned = mergedDf[~cond]

# Print the Dataset
print(mergedDfCleaned)
# print(len(mergedDfCleaned))

61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

            pcc          ba  htn 

## Part 3.4: Sort the dataset according to the Labels

In [ ]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sortedMergedDfCleaned = mergedDfCleaned.sort_values(by = 'Target').reset_index(drop = True)

# Print the dataset
print(mergedDfCleaned)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

            pcc          ba  htn   dm  cad appet   

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [ ]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dfDummies = pd.get_dummies(sortedMergedDfCleaned, columns = categorical_columns, prefix=categorical_columns)

# Print the dataset
print(dfDummies)
# print(dfDummies.columns.tolist()) # is this supposed to be in the final df? 

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [ ]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
def removeOutliers(df, num_cols):
    dfCopy = df.copy()
    for col in num_cols:
        mean = dfCopy[col].mean()
        std = dfCopy[col].std()
        upper = mean + 3 * std
        lower = mean - 3 * std
        dfCopy = dfCopy[(dfCopy[col] >= lower) & (dfCopy[col] <= upper)]
    return dfCopy
sortedMergedDfCleanedNoOutlier = removeOutliers(sortedMergedDfCleaned, numerical_columns)

# Print the dataset
print(sortedMergedDfCleanedNoOutlier)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

            pcc          ba  htn   dm  cad appet   

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [ ]:
# Normalize the all Numerical Attributes in the dataset.
def normalize(df, num_col):
    dfCopy = df.copy()
    for col in num_col:
        mean = dfCopy[col].mean()
        std = dfCopy[col].std()
        dfCopy[col] = (dfCopy[col] - mean) / std
    return dfCopy
sortedMergedDfCleanedNoOutlierNorm = normalize(sortedMergedDfCleanedNoOutlier, numerical_columns)
# Print the dataset
print(sortedMergedDfCleanedNoOutlierNorm)

     unique_id       age        bp       bgr        bu        sc       sod  \
2       343710  0.807138  0.727802  1.142120  3.999185  1.683685 -0.956087   
5       397388  0.932840  2.765646  3.507298 -0.225144  1.213526  0.447053   
7       438182  1.561350  0.727802  2.910477  3.774884  2.153844 -0.429910   
8       474407 -1.706899  1.746724 -0.316775 -0.038227  0.206042 -2.710013   
9       475200  0.995691  2.765646  0.921075  0.485141  3.900150 -0.079124   
..         ...       ...       ...       ...       ...       ...       ...   
148     486397 -1.769750 -0.291121  0.036896  0.111307 -0.264117 -0.956087   
149     203466 -1.644048 -1.310043 -0.537820 -0.860662 -0.129786 -0.429910   
150     482105 -0.261328 -1.310043 -0.582029  0.186074 -0.599945 -0.429910   
151     228849  0.618586  0.727802 -0.471507  0.335608 -0.129786 -0.079124   
152     996076  0.555735 -1.310043 -0.360984  0.298224 -0.129786  1.674801   

          pot      hemo       pcv  ...         pcc          ba 

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id and missingVal columns can be removed. We have already cleaned the data, and we know that each row corresponds to a specific person. The number of useful data is now preserved and no longer need a unique_id to perform operations on the dataset. 

In [ ]:
#Remove that column
finDf = sortedMergedDfCleanedNoOutlierNorm.drop(columns = ['unique_id', 'missingVal'])

# Print the dataset
print(finDf)

          age        bp       bgr        bu        sc       sod       pot  \
2    0.807138  0.727802  1.142120  3.999185  1.683685 -0.956087  1.367853   
5    0.932840  2.765646  3.507298 -0.225144  1.213526  0.447053 -1.298281   
7    1.561350  0.727802  2.910477  3.774884  2.153844 -0.429910  2.308841   
8   -1.706899  1.746724 -0.316775 -0.038227  0.206042 -2.710013 -1.298281   
9    0.995691  2.765646  0.921075  0.485141  3.900150 -0.079124  0.426864   
..        ...       ...       ...       ...       ...       ...       ...   
148 -1.769750 -0.291121  0.036896  0.111307 -0.264117 -0.956087 -0.827787   
149 -1.644048 -1.310043 -0.537820 -0.860662 -0.129786 -0.429910 -0.043630   
150 -0.261328 -1.310043 -0.582029  0.186074 -0.599945 -0.429910 -0.200461   
151  0.618586  0.727802 -0.471507  0.335608 -0.129786 -0.079124 -1.298281   
152  0.555735 -1.310043 -0.360984  0.298224 -0.129786  1.674801  0.583696   

         hemo       pcv      wbcc  ...        pc         pcc          ba  h

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [11]:
# Export the dataframe to a new csv file
finDf.to_csv('cleanedData.csv', index = False)

# Export the dataframe to a new json file
finDf.to_json('cleanedData.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here

| Age |    BP     |    BGR    |    BU     |    SC     |    SOD    |    POT    |    HEMO   |    PCV    |    WBCC   |    ...    |     PC    |     PCC       |     BA     |  HTN |   DM |   CAD |  APPET |   PE |  ANE | Target |
|-----|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|---------------|------------|------|------|-------|--------|------|------|--------|
|  2  | 0.807138  | 0.727802  | 1.142120  | 3.999185  | 1.683685  | -0.956087 | -2.902248 | -2.894560 | 0.517701  | ...       | abnormal   | notpresent    | notpresent  | yes  | yes  |   yes |  poor  |  yes |  yes |   ckd  |
|  5  | 0.932840  | 2.765646  | 3.507298  | -0.225144 | 1.213526  | 0.447053  | -0.604893 | -0.588449 | 0.782525  | ...       | normal     | notpresent    | present     | yes  | no   |   yes |  good  |  no  |  no  |   ckd  |
|  7  | 1.561350  | 0.727802  | 2.910477  | 3.774884  | 2.153844  | -0.429910 | -1.688551 | -1.597372 | -0.365045 | ...       | abnormal   | notpresent    | notpresent  | yes  | yes  |   yes |  good  |  no  |  no  |   ckd  |
|  8  | -1.706899 | 1.746724  | -0.316775 | -0.038227 | 0.206042  | -2.710013 | -2.642170 | -3.038692 | 1.930095  | ...       | abnormal   | present        | present     | no   | no   |   no  |  good  |  no  | yes  |   ckd  |
|  9  | 0.995691  | 2.765646  | 0.921075  | 0.485141  | 3.900150  | -0.079124 | -2.815556 | -2.606296 | -0.232633 | ...       | abnormal   | notpresent    | present     | yes  | yes  |   no  |  good  |  yes |  no  |   ckd  |
| 10  | 0.744287  | -1.310043 | 3.684134  | -0.187760 | 0.206042  | -1.833050 | -2.815556 | -2.750428 | 3.165939  | ...       | abnormal   | present        | notpresent  | yes  | no   |   no  |  poor  |  no  | yes  |   ckd  |
| 12  | 0.492884  | 1.746724  | 0.169523  | 2.466464  | 3.564322  | -1.657657 | -2.295400 | -2.173900 | -0.718144 | ...       | abnormal   | notpresent    | notpresent  | yes  | no   |   no  |  good  |  no  | no   |   ckd  |
| 13  | 1.309946  | -0.291121 | 2.048403  | 2.055246  | 3.295659  | -3.586975 | -2.165361 | -2.318032 | 1.532859  | ...       | abnormal   | present        | present     | yes  | yes  |   yes |  good  |  yes | yes  |   ckd  |
| 15  | -0.135626 | -1.310043 | 0.921075  | 1.905712  | 1.280691  | 0.096268  | -1.991975 | -2.318032 | 2.901115  | ...       | normal     | notpresent    | notpresent  | yes  | yes  |   no  |  good  |  no  | no   |   ckd  |
| 18  | 0.869989  | -0.291121 | 0.014792  | 0.036540  | 0.206042  | -0.780695 | -0.778278 | -0.732580 | -0.056084 | ...       | abnormal   | notpresent    | notpresent  | yes  | yes  |   no  |  good  |  no  | no   |   ckd  |
| 19  | 0.430033  | 0.727802  | 2.048403  | 1.195427  | 1.683685  | -0.605302 | -1.515166 | -1.453240 | -0.276771 | ...       | abnormal   | present        | present     | yes  | yes  |   no  |  good  |  yes | no   |   ckd  |
| 20  | 0.995691  | -1.310043 | 2.601015  | 0.634675  | 1.952347  | -0.605302 | -2.122014 | -2.173900 | -0.232633 | ...       | abnormal   | notpresent    | present     | yes  | yes  |   no  |  poor  |  yes | no   |   ckd  |
| 22  | 1.435648  | -1.310043 | -0.073626 | 3.139366  | 2.624003  | -0.780695 | -1.298434 | -1.309108 | 3.165939  | ...       | normal     | notpresent    | notpresent  | yes  | yes  |   no  |  poor  |  yes | no   |   ckd  |
| 23  | 1.435648  | -0.291121 | 2.158925  | 1.531878  | 1.482188  | -1.306872 | -1.731897 | -1.597372 | -1.071242 | ...       | abnormal   | present        | present     | yes  | yes  |   yes |  good  |  no  | no   |   ckd  |
| 29  | 0.681437  | 0.727802  | 4.015701  | -0.225144 | -0.062621 | -3.236190 | -1.731897 | -1.309108 | 1.268035  | ...       | normal     | notpresent    | notpresent  | no   | yes  |   no  |  poor  |  no  | no   |   ckd  |


In [ ]:
print(finDf.head(15)) # where are the other columns? 

         age        bp       bgr        bu        sc       sod       pot  \
2   0.807138  0.727802  1.142120  3.999185  1.683685 -0.956087  1.367853   
5   0.932840  2.765646  3.507298 -0.225144  1.213526  0.447053 -1.298281   
7   1.561350  0.727802  2.910477  3.774884  2.153844 -0.429910  2.308841   
8  -1.706899  1.746724 -0.316775 -0.038227  0.206042 -2.710013 -1.298281   
9   0.995691  2.765646  0.921075  0.485141  3.900150 -0.079124  0.426864   
10  0.744287 -1.310043  3.684134 -0.187760  0.206042 -1.833050 -2.082438   
12  0.492884  1.746724  0.169523  2.466464  3.564322 -1.657657  0.740527   
13  1.309946 -0.291121  2.048403  2.055246  3.295659 -3.586975 -0.670955   
15 -0.135626 -1.310043  0.921075  1.905712  1.280691  0.096268 -0.514124   
18  0.869989 -0.291121  0.014792  0.036540  0.206042 -0.780695  0.583696   
19  0.430033  0.727802  2.048403  1.195427  1.683685 -0.605302  0.897359   
20  0.995691 -1.310043  2.601015  0.634675  1.952347 -0.605302  1.681516   
22  1.435648

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: Cross reference LLM conversions with original data, make sure directions given to LLM is specific, and avoid giving large chunks of data at once (cut down to multiple chunks).

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [ ]:
# Code to rename all the columns in the dataset
rename_dict = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns
finDf.rename(columns=rename_dict, inplace=True)

# Display the modified DataFrame
print(finDf)

          age  blood pressure  blood glucose random  blood urea  \
2    0.807138        0.727802              1.142120    3.999185   
5    0.932840        2.765646              3.507298   -0.225144   
7    1.561350        0.727802              2.910477    3.774884   
8   -1.706899        1.746724             -0.316775   -0.038227   
9    0.995691        2.765646              0.921075    0.485141   
..        ...             ...                   ...         ...   
148 -1.769750       -0.291121              0.036896    0.111307   
149 -1.644048       -1.310043             -0.537820   -0.860662   
150 -0.261328       -1.310043             -0.582029    0.186074   
151  0.618586        0.727802             -0.471507    0.335608   
152  0.555735       -1.310043             -0.360984    0.298224   

     serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
2            1.683685 -0.956087   1.367853   -2.902248           -2.894560   
5            1.213526  0.447053  -1.298

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [ ]:
# Converted SQL to Pandas code
result = finDf.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)

   Target  count
0     ckd     22
1  notckd    111


Goes to the Target columns of the cleaned dataset (since we used dummies), counts how many entries are ckd and how many entries are notckd, then stores these numbers under a newly created column called count. The printed result returns two columns: Target and count. 